File: lib/src/models/download_configuration.dart
import 'package:pl_image_downloader/src/models/download_dictionary.dart';

import '../enum/download_mode.dart';
import '../enum/mime_type.dart';
import 'notification_config.dart';

/// Download Configuration
/// This class is used to store the download configuration.
class DownloadConfiguration {
  /// The name of the download file.
  final String? saveName;

  /// Whether to save the download file to external storage.
  final bool isExternalStorage; // Optional for android

  /// The download mode.
  final DownloadMode downloadMode;

  /// The mime type of the download file.
  final MimeType mimeType;

  /// The notification config.
  final NotificationConfig? notificationConfig;

  /// The retry count.
  final int retryCount;

  /// The download directory.
  final DownloadDirectory downloadDirectory;

  DownloadConfiguration({
    this.saveName,
    this.notificationConfig,
    this.isExternalStorage = false,
    this.downloadMode = DownloadMode.normal,
    this.mimeType = MimeType.imageJpeg,
    this.retryCount = 3,
    this.downloadDirectory = DownloadDirectory.downloads,
  });

  /// Convert the download configuration to a JSON object.
  Map<String, dynamic> toJson() {
    return {
      'saveName': saveName,
      'mimeType': mimeType.name,
      'downloadMode': downloadMode.name,
      'isExternalStorage': isExternalStorage,
      'retryCount': retryCount,
      'notificationConfig': notificationConfig?.toJson(),
      'downloadDirectory': downloadDirectory.directoryName,
    };
  }
}

File: android/src/main/kotlin/com/example/pl_image_downloader/models/DownloadConfiguration.kt
package com.example.pl_image_downloader.models

import com.example.pl_image_downloader.models.enum.DownloadDirectory
import com.example.pl_image_downloader.models.enum.DownloadMode
import com.example.pl_image_downloader.models.enum.MimeTypes
import com.example.pl_image_downloader.notification.NotificationConfig

/**
 * DownloadConfiguration
 * A data class representing the configuration for a download operation.
 */
data class DownloadConfiguration(
    val saveName: String? = null,
    val isExternalStorage: Boolean = false,
    val downloadMode: DownloadMode = DownloadMode.NORMAL,
    val mimeType: MimeTypes = MimeTypes.IMAGE_JPEG,
    val retryCount: Int = 3,
    val downloadDirectory: DownloadDirectory = DownloadDirectory.DOWNLOADS,
    val notificationConfig: NotificationConfig? = null,
) {
    companion object {

        /** * Provides a default DownloadConfiguration instance. */
        fun default(): DownloadConfiguration {
            return DownloadConfiguration()
        }

        /** * Checks if the current download mode. */
        fun isNormalMode(): Boolean = DownloadConfiguration().downloadMode == DownloadMode.NORMAL
        fun isBackgroundServiceMode(): Boolean = DownloadConfiguration().downloadMode == DownloadMode.RUNNING_BACKGROUND_SERVICE
    }
}


/** * Extension function to convert a Map to DownloadConfiguration. */
fun Map<*, *>.fromDownloadConfiguration(): DownloadConfiguration {
    return DownloadConfiguration(
        saveName = this["saveName"] as String?,
        isExternalStorage = this["isExternalStorage"] as Boolean,
        downloadMode = DownloadMode.fromValue(this["downloadMode"] as String),
        mimeType = MimeTypes.fromValue(this["mimeType"] as String),
        notificationConfig = null, // Handle NotificationConfig conversion if needed
        retryCount = this["retryCount"] as Int,
        downloadDirectory = DownloadDirectory.fromString(this["downloadDirectory"] as String)
    )
}